### PR TITLE
[14.0][IMP] purchase_sale_inter_company: lock incoming picking validation if intercompany transfer

### PR DESCRIPTION
**Pull request by toita86**

### CODE REVIEW

**Review by SicurSam — APPROVED** [review]

File: purchase_sale_inter_company/models/res_company.py [hunk 1]
```diff
@@ -36,3 +36,7 @@ class ResCompany(models.Model):
         help="Sync the receipt from the destination company with the "
         "delivery from the source company",
     )
+    block_po_manual_picking_validation = fields.Boolean(
+        string="Block manual validation of picking in the destination company",
+        help="Block manual validation of picking in the destination company",
```
Comment (PicchiSeba) [inline purchase_sale_inter_company/models/res_company.py]: **nitpick**: The help attribute is used to display additional information to the user.
Change this string to add information or remove it if you deem it necessary

File: purchase_sale_inter_company/models/res_config.py [hunk 2]
```diff
@@ -47,3 +47,9 @@ class InterCompanyRulesConfig(models.TransientModel):
         "the delivery from the source company",
         readonly=False,
     )
+    block_po_manual_picking_validation = fields.Boolean(
+        related="company_id.block_po_manual_picking_validation",
+        string="Block manual validation of picking in the destination company",
+        help="Block manual validation of picking in the destination company",
```
Comment (PicchiSeba) [inline purchase_sale_inter_company/models/res_config.py]: **nitpick**: unless you're redefining these attributes I would omit them. A field related to another one gets its values.

File: purchase_sale_inter_company/models/stock_picking.py [hunk 3]
```diff
@@ -106,6 +106,23 @@ def button_validate(self):
...
+        # if the flag is set, block the validation of the picking in the destination company
+        if self.env.company.block_po_manual_picking_validation:
+            for record in self.sudo():
+                dest_company = record.partner_id.commercial_partner_id.ref_company_ids
+                if (
+                    dest_company
+                    and record.picking_type_code == "incoming"
+                    and record.state == "assigned"
+                    or record.state == "done"
+                    or record.state == "waiting"
```
Comment (PicchiSeba) [inline purchase_sale_inter_company/models/stock_picking.py]: **question**: If I'm understanding this correctly you mean `(dest_company AND picking_type_code AND state == "assigned") OR state == "done" OR state == "waiting"`

And not `(dest_company AND picking_type_code) AND state == ("assigned" OR "done" OR "waiting")`.
Right?


**suggestion**: If so I would suggest to add brackets to help visualize the logical expression

```suggestion
                    (
                        dest_company
                        and record.picking_type_code == "incoming"
                        and record.state == "assigned"
                    )
                    or record.state in ["done", "waiting"]
```

File: purchase_sale_inter_company/models/stock_picking.py [hunk 4]
```diff
@@ -106,6 +106,23 @@ def button_validate(self):
                         dest_company,
                         record.sale_id,
                     )
+        # if the flag is set, block the validation of the picking in the destination company
+        if self.env.company.block_po_manual_picking_validation:
+            for record in self.sudo():
```
Comment (PicchiSeba) [inline purchase_sale_inter_company/models/stock_picking.py]: **question**: what would happen if we were to remove this `sudo()`?

Would we get an access error of some kind?

**Review by PicchiSeba — COMMENTED** [review]
Code review: a couple of questions and a few non-blocking fixes

File: purchase_sale_inter_company/models/stock_picking.py (hunk 3 above)
Comment (toita86) [reply purchase_sale_inter_company/models/stock_picking.py]: My intent was the second one, I have rewritten the expression based on your suggestion.

File: purchase_sale_inter_company/models/stock_picking.py (hunk 4 above)
Comment (toita86) [reply purchase_sale_inter_company/models/stock_picking.py]: Thank you for pointing this out, there are no access errors so it was just useless

**Review by PicchiSeba — APPROVED** [review]
Code review: LGTM

**Comment by francesco-ooops** [comment]
@pedrobaeza merge?

Rationale for this change: do not allow purchase picking to be validated before corresponding intercompany sale picking is validated

**Comment by pedrobaeza** [comment]
I don't get this need. Isn't the sales order auto-confirmed when the PO is confirmed?

**Comment by francesco-ooops** [comment]
@pedrobaeza the sale order is confirmed but the sale order picking is not (rightfully)

so what we're doing here is to prevent Purchasing company from receiving products that Selling company hasn't shipped

**Comment by pedrobaeza** [comment]
OK, understood.

/ocabot merge minor

**Pull request merged**